File: c/hanoi/hanoi.c
#include <stdio.h>
#include "hanoi.h"

extern struct stacks s;

int moves = 0;

void print_stacks(void) {
    printf(" { A => [");
    int i;

    // Print stack A
    for (i = 0; i < s.A_top; i++) {
        printf(" %d", *(s.A + i));
    }
    printf(" ], B => [");

    // Print stack B
    for (i = 0; i < s.B_top; i++) {
        printf(" %d", *(s.B + i));
    }
    printf(" ], C => [");

    // Print stack C
    for (i = 0; i < s.C_top; i++) {
        printf(" %d", *(s.C + i));
    }
    printf(" ] }\n");
}

char stacks_to_char(enum stack st) {
    if (st == A) return 'A';
    if (st == B) return 'B';
    return 'C';
}

void log_move(enum stack src, enum stack dst) {
    print_stacks();
    moves++;
    printf("Move #%d: %c -> %c\n", moves, stacks_to_char(src), stacks_to_char(dst));
}

void hanoi(int n, enum stack src, enum stack dst) {
    if (n == 1) {
        log_move(src, dst);
        make_move(src, dst);
    } else {
        // Determine the "via" stack for the final move of this round
        enum stack via = (A + B + C) - src - dst;

        hanoi(n - 1, src, via);
        log_move(src, dst);
        make_move(src, dst);
        hanoi(n - 1, via, dst);
    }
}

File: go/hanoi/hanoi.go
package main

import (
	"flag"
	"fmt"
)

const (
	A = iota
	B
	C
)

type stack struct {
	Name  string
	Discs []int
}

func (s *stack) pop() int {
	stack_length := len(s.Discs)
	disc := s.Discs[stack_length-1]
	s.Discs = s.Discs[0 : stack_length-1]
	return disc
}

func (s *stack) push(disc int) {
	s.Discs = append(s.Discs, disc)
}

var (
	discs  = flag.Int("discs", 7, "Number of discs to use.")
	moves  = 0
	stacks []*stack
)

func init() {
	flag.Parse()
}

func main() {
	fmt.Printf("Solving Towers of Hanoi for %d discs\n", *discs)

	// Create the stacks
	stacks = make([]*stack, 3, 3)
	stacks[A] = &stack{"A", make([]int, 0, *discs)}
	stacks[B] = &stack{"B", make([]int, 0, *discs)}
	stacks[C] = &stack{"C", make([]int, 0, *discs)}

	// Create the discs
	for i := 0; i < *discs; i++ {
		stacks[A].Discs = append(stacks[A].Discs, i)
	}

	hanoi(*discs, A, C)
}

func via_stack(src, dst int) int {
	return A + B + C - src - dst
}

func hanoi(d, src, dst int) {
	via := via_stack(src, dst)
	if d > 1 {
		hanoi(d-1, src, via)
	}

	moves++
	stacks[dst].push(stacks[src].pop())
	fmt.Printf("Move #%d: %s -> %s\n", moves, stacks[src].Name, stacks[dst].Name)

	if d > 1 {
		hanoi(d-1, via, dst)
	}
}
